### PR TITLE
[Refactor] Use MODELS registry in `mmengine`, delete basemodule

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

MMCV uses `MODELS` registry defined in mmengine to register convolution, padding, upsample norm layers .etc.
The following registry will be removed in mmcv:
-  `CONV_LAYERS`
- `NORM_LAYERS`
- `ACTIVATION_LAYERS`
- `PADDING_LAYERS`
- `UPSAMPLE_LAYERS`
- `PLUGIN_LAYERS`
- `DROPOUT_LAYERS`
- `POSITIONAL_ENCODING`
- `ATTENTION`
- `FEEDFORWARD_NETWORK`
- `TRANSFORMER_LAYER`
- `TRANSFORMER_LAYER_SEQUENCE`

All registries mentioned above will be removed,please use `mmegine.MODELS` to register your module instead.

The original usage:
```python
from mmcv.cnn import CONV_LAYERS
import torch.nn as nn


@CONV_LAYERS.register_module(nn.Module)
class CustomModule:
    ...
```
**Now we should use `MODELS` like this:**

```python
from mmengine import MODELS


@MODELS.register_module(nn.Module)
class CustomModule:
    ...
```
## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/cnn/bricks/padding.py [hunk 1]
```diff
@@ -2,12 +2,11 @@
...
 
 import torch.nn as nn
+from mmengine.registry import MODELS
 
-from .registry import PADDING_LAYERS
-
-PADDING_LAYERS.register_module('zero', module=nn.ZeroPad2d)
-PADDING_LAYERS.register_module('reflect', module=nn.ReflectionPad2d)
-PADDING_LAYERS.register_module('replicate', module=nn.ReplicationPad2d)
+MODELS.register_module('zero', module=nn.ZeroPad2d)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/padding.py:7]: The names of the modules are too simple. Maybe we can add some name mapping in the build function, and use the full name in the registry.

File: mmcv/cnn/builder.py [hunk 2]
```diff
@@ -1,30 +0,0 @@
-# Copyright (c) OpenMMLab. All rights reserved.
-from ..runner import Sequential
-from ..utils import Registry, build_from_cfg
-
-
-def build_model_from_cfg(cfg, registry, default_args=None):
```
Comment (RangiLyu) [inline mmcv/cnn/builder.py:6]: should avoid bc-breaking here

File: mmcv/cnn/bricks/registry.py [hunk 3]
```diff
@@ -1,16 +0,0 @@
-# Copyright (c) OpenMMLab. All rights reserved.
-from mmcv.utils import Registry
-
-CONV_LAYERS = Registry('conv layer')
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/registry.py:4]: avoid bc-breaking:
```
CONV_LAYERS = MODELS
......
```

File: mmcv/cnn/bricks/conv.py [hunk 4]
```diff
@@ -36,10 +35,10 @@ def build_conv_layer(cfg: Optional[Dict], *args, **kwargs) -> nn.Module:
         cfg_ = cfg.copy()
 
     layer_type = cfg_.pop('type')
-    if layer_type not in CONV_LAYERS:
+    if layer_type not in MODELS:
         raise KeyError(f'Unrecognized layer type {layer_type}')
     else:
-        conv_layer = CONV_LAYERS.get(layer_type)
+        conv_layer = MODELS.get(layer_type)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/conv.py:41]: the `get` method can not read the default_scope

File: mmcv/cnn/bricks/activation.py [hunk 5]
```diff
@@ -92,4 +92,4 @@ def build_activation_layer(cfg: Dict) -> nn.Module:
     Returns:
         nn.Module: Created activation layer.
     """
-    return build_from_cfg(cfg, ACTIVATION_LAYERS)
+    return build_from_cfg(cfg, MODELS)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/activation.py:95]:
```suggestion
    return MODELS.build(cfg)
```

File: mmcv/cnn/bricks/drop.py [hunk 6]
```diff
@@ -66,4 +65,4 @@ def __init__(self, drop_prob: float = 0.5, inplace: bool = False):
 
 def build_dropout(cfg: Dict, default_args: Optional[Dict] = None) -> Any:
     """Builder for drop out layers."""
-    return build_from_cfg(cfg, DROPOUT_LAYERS, default_args)
+    return build_from_cfg(cfg, MODELS, default_args)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/drop.py:68]: MODELS.build()

File: mmcv/cnn/bricks/norm.py [hunk 7]
```diff
@@ -97,10 +97,10 @@ def build_norm_layer(cfg: Dict,
     cfg_ = cfg.copy()
 
     layer_type = cfg_.pop('type')
-    if layer_type not in NORM_LAYERS:
+    if layer_type not in MODELS:
         raise KeyError(f'Unrecognized norm type {layer_type}')
 
-    norm_layer = NORM_LAYERS.get(layer_type)
+    norm_layer = MODELS.get(layer_type)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/norm.py:103]: default_scope problem

File: mmcv/cnn/bricks/padding.py [hunk 8]
```diff
@@ -28,10 +27,10 @@ def build_padding_layer(cfg: Dict, *args, **kwargs) -> nn.Module:
 
     cfg_ = cfg.copy()
     padding_type = cfg_.pop('type')
-    if padding_type not in PADDING_LAYERS:
+    if padding_type not in MODELS:
         raise KeyError(f'Unrecognized padding type {padding_type}.')
     else:
-        padding_layer = PADDING_LAYERS.get(padding_type)
+        padding_layer = MODELS.get(padding_type)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/padding.py:33]: also default_scope problem

File: mmcv/cnn/bricks/plugin.py [hunk 9]
```diff
@@ -80,10 +79,10 @@ def build_plugin_layer(cfg: Dict,
     cfg_ = cfg.copy()
 
     layer_type = cfg_.pop('type')
-    if layer_type not in PLUGIN_LAYERS:
+    if layer_type not in MODELS:
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/plugin.py:82]: also default_scope problem

File: mmcv/cnn/bricks/transformer.py [hunk 10]
```diff
@@ -37,27 +36,27 @@
 
 def build_positional_encoding(cfg, default_args=None):
     """Builder for Position Encoding."""
-    return build_from_cfg(cfg, POSITIONAL_ENCODING, default_args)
+    return build_from_cfg(cfg, MODELS, default_args)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/transformer.py:39]: MODELS.build()

File: mmcv/cnn/bricks/upsample.py [hunk 11]
```diff
@@ -76,10 +75,10 @@ def build_upsample_layer(cfg: Dict, *args, **kwargs) -> nn.Module:
     cfg_ = cfg.copy()
 
     layer_type = cfg_.pop('type')
-    if layer_type not in UPSAMPLE_LAYERS:
+    if layer_type not in MODELS:
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/upsample.py:78]: Also default_scope problem

File: mmcv/cnn/bricks/conv.py [hunk 12]
```diff
@@ -36,11 +35,11 @@ def build_conv_layer(cfg: Optional[Dict], *args, **kwargs) -> nn.Module:
         cfg_ = cfg.copy()
 
     layer_type = cfg_.pop('type')
-    if layer_type not in CONV_LAYERS:
-        raise KeyError(f'Unrecognized layer type {layer_type}')
-    else:
-        conv_layer = CONV_LAYERS.get(layer_type)
-
+    with MODELS.switch_scope_and_registry(None) as TARGET_MODELS:
+        conv_layer = TARGET_MODELS.get(layer_type)
```
Comment (RangiLyu) [inline mmcv/cnn/bricks/conv.py:39]:
```suggestion
    with MODELS.switch_scope_and_registry(None) as registry:
        conv_layer = registry.get(layer_type)
```
Also add some commnets

**Comment by gaotongxiao** [comment]
@ZwwWayne MMDet needs to be updated so that we can run full tests. Ref https://github.com/open-mmlab/mmocr/pull/1261/files

File: mmcv/ops/carafe.py [hunk 13]
```diff
@@ -4,11 +4,12 @@
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
+from mmengine import MODELS
```
Comment (zhouzaida) [inline mmcv/ops/carafe.py:7]:
```suggestion
from mmengine.registry import MODELS
```

File: tests/test_cnn/test_build_layers.py [hunk 14]
```diff
@@ -172,6 +176,21 @@ def test_build_norm_layer():
 
 
 def test_build_activation_layer():
+    all_act_modules = [
```
Comment (zhouzaida) [inline tests/test_cnn/test_build_layers.py:179]:
```suggestion
    act_names = [
```

File: tests/test_cnn/test_build_layers.py [hunk 15]
```diff
@@ -172,6 +176,21 @@ def test_build_norm_layer():
...
+    custom_act_module = dict()
+    for module_name in ['activation', 'hsigmoid', 'hswish', 'swish']:
+        _act_module = import_module(f'mmcv.cnn.bricks.{module_name}')
+        _act_module = {
+            key: value
+            for key, value in _act_module.__dict__.items()
+            if isinstance(value, type) and issubclass(value, nn.Module)
+        }
+        custom_act_module.update(_act_module)
+    all_act_modules += list(custom_act_module.keys())
```
Comment (zhouzaida) [inline tests/test_cnn/test_build_layers.py:192]:
```suggestion
    for module_name in ['activation', 'hsigmoid', 'hswish', 'swish']:
        act_module = import_module(f'mmcv.cnn.bricks.{module_name}')
        for key, value in act_module.__dict__.items():
            if isinstance(value, type) and issubclass(value, nn.Module):
                act_names.append(key)
```

File: tests/test_cnn/test_build_layers.py [hunk 16]
```diff
@@ -207,6 +227,18 @@ def test_build_activation_layer():
...
+    custom_pad_modules = dict()
+    for module_name in ['padding']:
+        _pad_module = import_module(f'mmcv.cnn.bricks.{module_name}')
+        _pad_module = {
+            key: value
+            for key, value in _pad_module.__dict__.items()
+            if isinstance(value, type) and issubclass(value, nn.Module)
+        }
+        custom_pad_modules.update(_pad_module)
+    all_pad_modules += list(custom_pad_modules)
```
Comment (zhouzaida) [inline tests/test_cnn/test_build_layers.py:240]: same as above

File: mmcv/cnn/bricks/conv.py [hunk 17]
```diff
@@ -36,11 +35,15 @@ def build_conv_layer(cfg: Optional[Dict], *args, **kwargs) -> nn.Module:
...
     layer_type = cfg_.pop('type')
-    if layer_type not in CONV_LAYERS:
-        raise KeyError(f'Unrecognized layer type {layer_type}')
-    else:
-        conv_layer = CONV_LAYERS.get(layer_type)
 
+    # Switch registry to the target scope. If `conv_layer` cannot be found
+    # in the registry, fallback to search `conv_layer` in the
+    # mmengine.MODELS.
+    with MODELS.switch_scope_and_registry(None) as registry:
```
[the remark this answers is not in the document]
Comment (HAOCHENYE) [reply mmcv/cnn/bricks/conv.py:42]: `None` should be allowed, I'll modify the type hint of `switch_scope_and_registry`

File: docs/en/understand_mmcv/registry.md [hunk 18]
```diff
@@ -129,9 +129,9 @@ Basically, there are two ways to build a module from child or sibling registries
    In MMDetection we define:
 
    ```python
-   from mmcv.utils import Registry
-   from mmcv.cnn import MODELS as MMCV_MODELS
-   MODELS = Registry('model', parent=MMCV_MODELS)
+   from mmcv.registry import Registry
```
Comment (zhouzaida) [inline docs/en/understand_mmcv/registry.md:132]:
```suggestion
   from mmengine.registry import Registry
```

File: docs/zh_cn/understand_mmcv/registry.md [hunk 19]
```diff
@@ -139,9 +139,9 @@ CONVERTERS = Registry('converter', build_func=build_converter)
    我们在 MMClassification 中定义：
 
    ```python
-   from mmcv.utils import Registry
-   from mmcv.cnn import MODELS as MMCV_MODELS
-   MODELS = Registry('model', parent=MMCV_MODELS)
+   from mmcv.registry import Registry
+   from mmcv.registry import MODELS as MMENGINE_MODELS
```
Comment (zhouzaida) [inline docs/zh_cn/understand_mmcv/registry.md:143]:
```suggestion
   from mmengine.registry import Registry
   from mmengine.registry import MODELS as MMENGINE_MODELS
```

**Comment by zhouzaida** [comment]
`tests/test_runner/test_basemodule.py` should also be removed.

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
